# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [150]:
import pandas as pd
import sqlite3 as s3
import warnings
warnings.simplefilter('ignore')

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [151]:
go_sales = s3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [152]:
product = pd.read_sql_query('SELECT * FROM product', go_sales)
product_type = pd.read_sql_query('SELECT * FROM product_type', go_sales)
product_line = pd.read_sql_query('SELECT * FROM product_line', go_sales)
sales_staff = pd.read_sql_query('SELECT * FROM sales_staff', go_sales)
sales_branch = pd.read_sql_query('SELECT * FROM sales_branch', go_sales)
retailer_site = pd.read_sql_query('SELECT * FROM retailer_site', go_sales)
country = pd.read_sql_query('SELECT * FROM country', go_sales)
order_header = pd.read_sql_query('SELECT * FROM order_header', go_sales)
order_details = pd.read_sql_query('SELECT * FROM order_details', go_sales)
sales_target = pd.read_sql_query('SELECT * FROM sales_targetData', go_sales)
returned_item = pd.read_sql_query('SELECT * FROM returned_item', go_sales)
return_reason = pd.read_sql_query('SELECT * FROM return_reason', go_sales)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [153]:
cost_range = (product['PRODUCTION_COST'].astype(float) < 100) & (product['PRODUCTION_COST'].astype(float) > 50)
pd.DataFrame(product.loc[cost_range, ['PRODUCT_NAME', 'PRODUCTION_COST']])

PRODUCT_NAME PRODUCTION_COST
13                Blue Steel Putter              60
14            Blue Steel Max Putter            81.8
17              Course Pro Golf Bag              80
24                  Hibernator Lite              60
25                       Hibernator              86
28  Hibernator Self - Inflating Mat           78.55
31              Hibernator Camp Cot           65.33
32     Canyon Mule Climber Backpack            62.5
56          Granite Climbing Helmet           52.86
58            Husky Harness Extreme              55
69                   Granite Hammer           57.13
71        TrailChef Deluxe Cook Set           85.11
81                        Polar Ice           73.33
83                     Polar Sports              80
84                       Polar Wave              60
85                    Polar Extreme            72.5
88                     Edge Extreme              80
91                        Seeker 35           79.19
92                        Seeker 50           92.58
93           TrailChef Double Flame              75
94                   Seeker Extreme           94.12
97                   Glacier Deluxe           56.25
98                      Glacier GPS           78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [154]:
margin_range = (product['MARGIN'].astype(float) < .20) | (product['MARGIN'].astype(float) > .60)
pd.DataFrame(product.loc[margin_range, ['PRODUCT_NAME', 'MARGIN']])

PRODUCT_NAME MARGIN
29                 Hibernator Pad    .17
32   Canyon Mule Climber Backpack    .17
100             BugShield Natural    .67
101               BugShield Spray    .67
102         BugShield Lotion Lite     .7
103              BugShield Lotion    .63
105             BugShield Extreme    .63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [155]:
currency = country['CURRENCY_NAME'] == 'francs'
pd.DataFrame(country.loc[currency, ['COUNTRY']].sort_values(by=['COUNTRY']))

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [156]:
filter = product['MARGIN'].astype(float) > .50
pd.DataFrame(product.loc[filter, ['INTRODUCTION_DATE']].drop_duplicates())

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [157]:
filter = (sales_branch['ADDRESS2'].notna()) & (sales_branch['REGION'].notna())
pd.DataFrame(sales_branch.loc[filter, ['ADDRESS1', 'CITY']])

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [158]:
currency = (country['CURRENCY_NAME'] == 'new dollar') | (country['CURRENCY_NAME'] == 'dollars')
pd.DataFrame(country.loc[currency, ['COUNTRY']].sort_values(by=['COUNTRY']))

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [159]:
filter = (retailer_site['POSTAL_ZONE'].str.startswith('D')) & (retailer_site['ADDRESS2'].notna())
pd.DataFrame(retailer_site.loc[filter, ['ADDRESS1', 'ADDRESS2', 'CITY']])

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [160]:
len(returned_item)

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [161]:
filter = retailer_site[retailer_site['REGION'].notna()]
filter['REGION'].drop_duplicates().count()

110

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [162]:
pd.DataFrame({ 'laagste' : [product['MARGIN'].astype(float).min()], 'hoogste' : [product['MARGIN'].astype(float).max()], 'gemiddelde' : [product['MARGIN'].astype(float).mean()] })

laagste  hoogste  gemiddelde
0     0.17      0.7    0.396609

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [163]:
retailer_site['ADDRESS2'].isna().sum()

286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [164]:
filter = order_details['UNIT_SALE_PRICE'].astype(float) < order_details['UNIT_PRICE'].astype(float)
filtered = pd.DataFrame(order_details.loc[filter, :])
filtered['UNIT_COST'].astype(float).mean()

125.42305809556548

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [165]:
filter = sales_staff['POSITION_EN'].value_counts().reset_index()
filter.columns = ['POSITION_EN', 'COUNT']
filter

POSITION_EN  COUNT
0  Level 2 Sales Representative     39
1  Level 3 Sales Representative     24
2                Branch Manager     19
3  Level 1 Sales Representative     13
4              District Manager      4
5              Regional Manager      2
6               General Manager      1

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [166]:
phone_counts = sales_staff['WORK_PHONE'].value_counts()
filter = phone_counts > 4
result = phone_counts[filter].reset_index()
result.columns = ['WORK_PHONE', 'COUNT']
result

WORK_PHONE  COUNT
0     1 (305) 557-4810      6
1     1 (617) 268-6754      6
2  +31 (0)20 692 93 94      5
3     1 (403) 232-5986      5
4  +(61) 03 2982  4242      5
5    +(44) 121 3505267      5
6     1 (310) 281-5722      5
7    +33 1 68 94 52 20      5
8     1 (206) 292-0012      5
9     1 (416) 493-5595      5

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [167]:
merged = pd.merge(retailer_site, country, left_on = 'COUNTRY_CODE', how = 'inner', right_on = 'COUNTRY_CODE')
filter = merged['COUNTRY'] == 'Netherlands'
pd.DataFrame(retailer_site.loc[filter, ['ADDRESS1', 'CITY']])

ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [168]:
merged = pd.merge(product, product_type, left_on = 'PRODUCT_TYPE_CODE', how = 'inner', right_on = 'PRODUCT_TYPE_CODE')
filter = merged['PRODUCT_TYPE_EN'] == 'Eyewear'
pd.DataFrame(merged.loc[filter, ['PRODUCT_NAME']])

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [169]:
merged = pd.merge(sales_staff, sales_branch, left_on='SALES_BRANCH_CODE', how='inner', right_on='SALES_BRANCH_CODE')
filter = merged['POSITION_EN'] == 'Branch Manager'
pd.DataFrame(merged.loc[filter, ['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']])

ADDRESS1 FIRST_NAME      LAST_NAME
7                     Victoria House     Walter         Taylor
11                    3 Albany Court      Polly         Linton
19                 Interleuvenlaan 2      Gracy        Gellens
27                    Schwabentor 35      Frank          Fuchs
30                  Leopoldstraße 36      Fritz         Hirsch
33                   Piazza Duomo, 1      Maria      Iacobucci
38               Singelgravenplein 4       Kick        Kalkman
43               Isafjordsgatan 30 C      Karin      Bergström
44     75, rue du Faubourg St-Honoré      Denis           Pagé
47      7800, 756 - 6th Avenue. S.W.      Sally          White
50                  789 Yonge Street    Georges  Saint-Germain
56            1288 Dorchester Avenue       Alex      Rodriguez
60                   299 Yale Avenue   Samantha          Floyd
65                    10032 NW 186th      Julie          Olsen
69        1288 South Barrington Ave.      Corey         Wright
80               6c, rue de l'Église      Julie        Béchard
84  Prol. Paseo de la Reforma No. 51      Lucía          Reyna
91                  2315 Queen's Ave      Alice         Walter
98             Avenida Paulista, 333     Bayard          Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [170]:
merged = pd.merge(sales_staff, order_header, left_on='SALES_STAFF_CODE', how='outer', right_on='SALES_STAFF_CODE')
filter = (merged['POSITION_EN'].str.contains('Manager')) #& (merged['ORDER_NUMBER'].notna())
pd.DataFrame(merged.loc[filter, ['POSITION_EN', 'ORDER_DATE']].drop_duplicates())

POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [171]:
product_merged = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE', how='inner', right_on='PRODUCT_TYPE_CODE')
merged = pd.merge(product_merged, order_details, left_on='PRODUCT_NUMBER', how='inner', right_on='PRODUCT_NUMBER')
filter = merged['QUANTITY'].astype(int) > 750
pd.DataFrame(merged.loc[filter, ['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates())

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [172]:
merged = pd.merge(product, order_details, left_on='PRODUCT_NUMBER', how='inner', right_on='PRODUCT_NUMBER')
filter = (merged['UNIT_PRICE'].astype(float) - merged['UNIT_SALE_PRICE'].astype(float)) / merged['UNIT_PRICE'].astype(float) > .40
pd.DataFrame(merged.loc[filter, ['PRODUCT_NAME']])
# merged

PRODUCT_NAME
2765   Hailstorm Titanium Woods Set
2972   Hailstorm Titanium Woods Set
37446             BugShield Natural
37514             BugShield Natural
37967               BugShield Spray
38306         BugShield Lotion Lite
38315         BugShield Lotion Lite
38318         BugShield Lotion Lite
38408         BugShield Lotion Lite
38434         BugShield Lotion Lite
38570         BugShield Lotion Lite
38623         BugShield Lotion Lite
38772              BugShield Lotion
38834              BugShield Lotion
38837              BugShield Lotion
39000              BugShield Lotion
39561             BugShield Extreme
39652             BugShield Extreme
39666             BugShield Extreme
39682             BugShield Extreme
39697             BugShield Extreme
39698             BugShield Extreme
39706             BugShield Extreme
39722             BugShield Extreme
39820             BugShield Extreme
39843             BugShield Extreme
39851             BugShield Extreme
40799                Sun Shelter 15
40803                Sun Shelter 15
40827                Sun Shelter 15
40993                Sun Shelter 15
41194                Sun Shelter 30
41254                Sun Shelter 30
41503                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [173]:
order_merged = pd.merge(order_details, returned_item, left_on='ORDER_DETAIL_CODE', how='inner', right_on='ORDER_DETAIL_CODE')
merged = pd.merge(order_merged, return_reason, left_on='RETURN_REASON_CODE', how='outer', right_on='RETURN_REASON_CODE')
filter = merged['RETURN_QUANTITY'].astype(float) / merged['QUANTITY'].astype(float) > .90
pd.DataFrame(merged.loc[filter, ['RETURN_DESCRIPTION_EN']])

RETURN_DESCRIPTION_EN
451   Wrong product ordered
452   Wrong product ordered
453   Wrong product ordered
454   Wrong product ordered
455   Wrong product ordered
..                      ...
701  Unsatisfactory product
702  Unsatisfactory product
703  Unsatisfactory product
704  Unsatisfactory product
705  Unsatisfactory product

[255 rows x 1 columns]

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [174]:
merged = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')
merged.groupby('PRODUCT_TYPE_EN').size().reset_index(name='PRODUCT_COUNT')

PRODUCT_TYPE_EN  PRODUCT_COUNT
0             Binoculars              4
1   Climbing Accessories              7
2           Cooking Gear             10
3                Eyewear              5
4              First Aid              5
5       Golf Accessories              4
6      Insect Repellents              5
7                  Irons              4
8                 Knives              5
9               Lanterns             12
10            Navigation              4
11                 Packs              6
12               Putters              3
13                  Rope              4
14                Safety              4
15         Sleeping Bags              7
16             Sunscreen              5
17                 Tents              6
18                 Tools              6
19               Watches              5
20                 Woods              4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [175]:
merged = pd.merge(retailer_site, country, on='COUNTRY_CODE')
merged.groupby('COUNTRY_CODE').size().reset_index(name='BRANCH_COUNT')

COUNTRY_CODE  BRANCH_COUNT
0             1            29
1            10            12
2            11            25
3            12            10
4            13            10
5            14            10
6            16            10
7            17            10
8            18             6
9            19             9
10            2            32
11           20            10
12           21             1
13           22             5
14            3           101
15            4            31
16            5            10
17            6            12
18            7            20
19            8            13
20            9            25

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [176]:
merged = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')
cooking_gear_data = merged[merged['PRODUCT_TYPE_EN'] == 'Cooking Gear']
order_details['UNIT_SALE_PRICE'] = order_details['UNIT_SALE_PRICE'].astype(float)
summary = order_details.groupby('PRODUCT_NUMBER').agg({'QUANTITY': 'sum', 'UNIT_SALE_PRICE': 'mean'}).reset_index()
cooking_gear_summary = pd.merge(cooking_gear_data, summary, left_on='PRODUCT_NUMBER', right_on='PRODUCT_NUMBER')
sorted_summary = cooking_gear_summary.groupby('PRODUCT_NAME').agg({'QUANTITY': 'sum', 'UNIT_SALE_PRICE': 'mean'}).sort_values(by='QUANTITY', ascending=False)
sorted_summary.reset_index(inplace=True)
sorted_summary

PRODUCT_NAME  QUANTITY  UNIT_SALE_PRICE
0        TrailChef Water Bag     41770         5.999266
1              TrailChef Cup     31834         6.945976
2          TrailChef Canteen     31712        11.930119
3      TrailChef Kitchen Kit     25964        22.210198
4         TrailChef Utensils     15928        17.132397
5     TrailChef Double Flame     14510       131.348013
6         TrailChef Cook Set     12700        51.204554
7     TrailChef Single Flame     10614        63.572397
8           TrailChef Kettle      9900        12.130783
9  TrailChef Deluxe Cook Set      3744       120.303667

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [177]:
sales_merged = pd.merge(sales_staff, sales_branch, on = 'SALES_BRANCH_CODE', how = 'inner')
merged = pd.merge(sales_merged, sales_target, on = 'SALES_STAFF_CODE', how = 'inner')
filter = merged[merged['CITY'] == 'Amsterdam']
resultaat = filter[filter['SALES_YEAR'] == '2022'].groupby(['FIRST_NAME', 'LAST_NAME'])['SALES_TARGET'].agg(['min', 'max'])
resultaat.index.names = ['Voornaam', 'Achternaam']
resultaat.reset_index(inplace=True)
resultaat

Voornaam        Achternaam   min   max
0  Belinda  Jansen-Velasquez  1005   963
1    Ellen           Shapiro  1004   986
2      Jan         Haverkamp  1048   993
3    Maria          Laponder  1000  9956

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [178]:
sales_merged = pd.merge(sales_branch, retailer_site, on = 'COUNTRY_CODE', how = 'inner')
merged = pd.merge(sales_merged, country, on = 'COUNTRY_CODE', how = 'inner')
filter = merged.groupby('COUNTRY')['CITY_y'].nunique().reset_index(name='aantal_steden')
resultaat = pd.merge(merged, filter, on='COUNTRY', how='left')
resultaat.rename(columns={'CITY_x': 'verkoper'}, inplace=True)
resultaat = resultaat[['COUNTRY', 'verkoper', 'aantal_steden']]
resultaat.drop_duplicates(inplace=True)
resultaat

COUNTRY          verkoper  aantal_steden
0           Germany           Hamburg             18
32          Germany           München             18
64           Sweden             Kista             10
76           Canada           Calgary             16
107          Canada           Toronto             16
138   United States            Boston             62
239   United States           Seattle             62
340   United States       Los Angeles             62
441   United States             Miami             62
542          France              Lyon             14
571          Mexico  Distrito Federal              6
581           Japan             Tokyo              5
606           Japan        Osaka City              5
631       Australia         Melbourne              6
641           Spain            Bilbao              5
650          Brazil         São Paulo              1
651         Finland            Kuopio              5
656           Korea             Seoul              6
666          Taiwan         San Chung              9
676           China          Shanghai              7
686  United Kingdom            London              9
711  United Kingdom        Birmingham              9
736     Switzerland            Zürich              9
749         Belgium          Heverlee              8
759         Austria              Wien              9
769          France             Paris             14
798           Italy            Milano              9
810     Netherlands         Amsterdam             15

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [179]:
niks_verkocht = ~sales_staff['SALES_STAFF_CODE'].isin(order_header['SALES_STAFF_CODE'])
sales_staff.loc[(niks_verkocht), ('FIRST_NAME', 'LAST_NAME')]

FIRST_NAME      LAST_NAME
7      Walter         Taylor
8    Penelope        Tamrine
11      Polly         Linton
19      Gracy        Gellens
26       Else         Mörike
27      Frank          Fuchs
30      Fritz         Hirsch
31       Jörg          Kunze
33      Maria      Iacobucci
38       Kick        Kalkman
43      Karin      Bergström
44      Denis           Pagé
47      Sally          White
49      Frank        Bretton
50    Georges  Saint-Germain
56       Alex      Rodriguez
60   Samantha          Floyd
65      Julie          Olsen
69      Corey         Wright
76    Estelle        Johnson
77       Greg        Belding
80      Julie        Béchard
84      Lucía          Reyna
89     Kazumi        Uragome
91      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [180]:
gemiddelde_margin = product['MARGIN'].astype(float).mean()
aantal_producten = 0
totale_margin = 0
for index, row in product.iterrows(): 
    if float(row['MARGIN']) < gemiddelde_margin:  
        aantal_producten += 1
        totale_margin += float(row['MARGIN'])
        
if aantal_producten != 0:
    gemiddelde_marge_lager_dan_gemiddelde = totale_margin / aantal_producten
else:
    gemiddelde_marge_lager_dan_gemiddelde = 0

overzicht = pd.DataFrame({
    'Aantal_producten_met_lagere_marge': [aantal_producten],
    'Gemiddelde_lagere_marge': [gemiddelde_marge_lager_dan_gemiddelde]
})
overzicht

Aantal_producten_met_lagere_marge  Gemiddelde_lagere_marge
0                                 59                 0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [181]:
merged_data = pd.merge(product, order_details, on = 'PRODUCT_NUMBER', how = 'inner')
meer_dan_500 = merged_data[merged_data['UNIT_SALE_PRICE'].astype(float) > 500]
nooit_teruggebracht = meer_dan_500[~meer_dan_500['ORDER_DETAIL_CODE'].isin(returned_item['ORDER_DETAIL_CODE'])]
grouped_nooit_teruggebracht = nooit_teruggebracht.groupby('PRODUCT_NAME').size()
pd.DataFrame(grouped_nooit_teruggebracht.index, columns=['PRODUCT_NAME'])

PRODUCT_NAME
0               Hailstorm Steel Irons
1           Hailstorm Steel Woods Set
2            Hailstorm Titanium Irons
3        Hailstorm Titanium Woods Set
4                      Husky Rope 200
5          Lady Hailstorm Steel Irons
6      Lady Hailstorm Steel Woods Set
7       Lady Hailstorm Titanium Irons
8   Lady Hailstorm Titanium Woods Set
9                           Star Dome
10                       Star Gazer 2
11                       Star Gazer 3
12                       Star Gazer 6

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [182]:
manager_status = []
for index, row in sales_staff.iterrows():
    if 'Manager' in row['POSITION_EN']:
                manager_status.append('Ja')
    else:
        manager_status.append('Nee')

sales_staff['IsManager'] = manager_status
sales_staff[['LAST_NAME', 'IsManager']]

LAST_NAME IsManager
0    Savolainen       Nee
1           Kim       Nee
2          Choi       Nee
3           Kao       Nee
4          Fang       Nee
..          ...       ...
97     Iglesias       Nee
98        Lopes        Ja
99       Barros       Nee
100   Guimarães       Nee
101    Lindholm       Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [183]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [184]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 